File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionAwarenessGroup.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.processors.platform.client.cache;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.apache.ignite.binary.BinaryRawWriter;
import org.apache.ignite.cache.CacheKeyConfiguration;
import org.apache.ignite.configuration.CacheConfiguration;
import org.apache.ignite.internal.processors.cache.DynamicCacheDescriptor;
import org.apache.ignite.internal.processors.cache.binary.CacheObjectBinaryProcessorImpl;
import org.apache.ignite.internal.processors.platform.client.ClientBitmaskFeature;
import org.apache.ignite.internal.processors.platform.client.ClientProtocolContext;
import org.jetbrains.annotations.Nullable;

/**
 * Partitions mapping associated with a group of caches. This group may contain caches from different cache groups,
 * the grouping criteria is - the same mapping and default or non-default affinity function flag.
 */
class ClientCachePartitionAwarenessGroup {
    /** Partition mapping. If {@code null} then cache must be excluded in partition awareness usage (e.g. REPLICATED cache).  */
    private final @Nullable ClientCachePartitionMapping mapping;

    /** {@code true} if the RendezvousAffinityFunction is used with the default affinity key mapper. */
    private final boolean dfltAffinity;

    /** Descriptor of the associated caches. */
    private final Map<Integer, CacheConfiguration<?, ?>> cacheCfgs = new HashMap<>();

    /**
     * @param mapping Partition mapping.
     * @param dfltAffinity {@code true} if the default affinity or a custom affinity mapper was used.
     */
    public ClientCachePartitionAwarenessGroup(@Nullable ClientCachePartitionMapping mapping, boolean dfltAffinity) {
        this.mapping = mapping;
        this.dfltAffinity = dfltAffinity;
    }

    /**
     * Write mapping using binary writer.
     *
     * @param proc Binary processor.
     * @param writer Binary Writer.
     * @param cpctx Protocol context.
     */
    public void write(CacheObjectBinaryProcessorImpl proc, BinaryRawWriter writer, ClientProtocolContext cpctx) {
        boolean isPartitionAwarenessApplicable = mapping != null;

        writer.writeBoolean(isPartitionAwarenessApplicable);

        writer.writeInt(cacheCfgs.size());

        if (isPartitionAwarenessApplicable) {
            for (Map.Entry<Integer, CacheConfiguration<?, ?>> entry: cacheCfgs.entrySet()) {
                writer.writeInt(entry.getKey());

                writeCacheKeyConfiguration(writer, proc, entry.getValue().getKeyConfiguration());
            }

            mapping.write(writer);

            if (cpctx.isFeatureSupported(ClientBitmaskFeature.ALL_AFFINITY_MAPPINGS))
                writer.writeBoolean(dfltAffinity);
        }
        else {
            for (int cacheId : cacheCfgs.keySet())
                writer.writeInt(cacheId);
        }
    }

    /**
     * Add caches to the same affinity group.
     * @param descs Cache descriptors.
     */
    public void addAll(List<DynamicCacheDescriptor> descs) {
        for (DynamicCacheDescriptor desc : descs)
            cacheCfgs.putIfAbsent(desc.cacheId(), desc.cacheConfiguration());
    }

    /** */
    private static void writeCacheKeyConfiguration(
        BinaryRawWriter writer,
        CacheObjectBinaryProcessorImpl binProc,
        CacheKeyConfiguration[] keyCfgs
    ) {
        if (keyCfgs == null) {
            writer.writeInt(0);

            return;
        }

        writer.writeInt(keyCfgs.length);

        for (CacheKeyConfiguration keyCfg : keyCfgs) {
            int keyTypeId = binProc.typeId(keyCfg.getTypeName());
            int affinityKeyFieldId = binProc.binaryContext().fieldId(keyTypeId, keyCfg.getAffinityKeyFieldName());

            writer.writeInt(keyTypeId);
            writer.writeInt(affinityKeyFieldId);
        }
    }

    /** {@inheritDoc} */
    @Override public boolean equals(Object o) {
        if (this == o)
            return true;

        if (o == null || getClass() != o.getClass())
            return false;

        ClientCachePartitionAwarenessGroup group = (ClientCachePartitionAwarenessGroup)o;

        return dfltAffinity == group.dfltAffinity && Objects.equals(mapping, group.mapping);
    }

    /** {@inheritDoc} */
    @Override public int hashCode() {
        return Objects.hash(mapping, dfltAffinity);
    }
}

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/ClientCachePartitionAwarenessGroup.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite.Core.Impl.Client.Cache
{
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using Apache.Ignite.Core.Impl.Binary;
    using Apache.Ignite.Core.Impl.Binary.IO;

    /// <summary>
    /// Partition mapping associated with the group of caches.
    /// Mirrors corresponding Java class.
    /// </summary>
    internal class ClientCachePartitionAwarenessGroup
    {
        /** */
        private readonly List<KeyValuePair<Guid, List<int>>> _partitionMap;

        /** */
        private readonly List<KeyValuePair<int, Dictionary<int, int>>> _caches;

        public ClientCachePartitionAwarenessGroup(IBinaryStream stream)
        {
            // Whether this group is eligible for client-side partition awareness.
            var applicable = stream.ReadBool();

            var cachesCount = stream.ReadInt();
            _caches = new List<KeyValuePair<int, Dictionary<int, int>>>(cachesCount);

            for (var i = 0; i < cachesCount; i++)
            {
                var cacheId = stream.ReadInt();
                if (!applicable)
                {
                    _caches.Add(new KeyValuePair<int, Dictionary<int, int>>(cacheId, null));
                    continue;
                }

                var keyCfgCount = stream.ReadInt();
                Dictionary<int, int> keyCfgs = null;
                if (keyCfgCount > 0)
                {
                    keyCfgs = new Dictionary<int, int>(keyCfgCount);
                    for (var j = 0; j < keyCfgCount; j++)
                    {
                        keyCfgs[stream.ReadInt()] = stream.ReadInt();
                    }
                }

                _caches.Add(new KeyValuePair<int, Dictionary<int, int>>(cacheId, keyCfgs));
            }

            if (!applicable)
                return;

            var partMapSize = stream.ReadInt();
            _partitionMap = new List<KeyValuePair<Guid, List<int>>>(partMapSize);

            var reader = BinaryUtils.Marshaller.StartUnmarshal(stream);

            for (var i = 0; i < partMapSize; i++)
            {
                var nodeId = reader.ReadGuid();
                Debug.Assert(nodeId != null);

                var partCount = stream.ReadInt();
                var parts = new List<int>(partCount);

                for (int j = 0; j < partCount; j++)
                {
                    parts.Add(stream.ReadInt());
                }

                _partitionMap.Add(new KeyValuePair<Guid, List<int>>(nodeId.Value, parts));
            }
        }

        /// <summary>
        /// Gets the caches.
        /// </summary>
        public ICollection<KeyValuePair<int, Dictionary<int, int>>> Caches
        {
            get { return _caches; }
        }

        /// <summary>
        /// Gets the partition map: node id -> partitions.
        /// </summary>
        public ICollection<KeyValuePair<Guid, List<int>>> PartitionMap
        {
            get { return _partitionMap; }
        }
    }
}
